sycl/include/sycl/ext/intel/experimental/device_architecture.hpp

```diff
@@ -0,0 +1,241 @@
+// the device compiler will set this variable to false, and that will trigger
+// an error for code that uses "if_architecture_is".
+static constexpr bool is_allowable_aot_mode =
+    (__SYCL_TARGET_INTEL_X86_64__ == 1) ||
+    (__SYCL_TARGET_INTEL_GPU_BDW__ == 1) ||
+    (__SYCL_TARGET_INTEL_GPU_SKL__ == 1) ||
+    (__SYCL_TARGET_INTEL_GPU_KBL__ == 1) ||
+    (__SYCL_TARGET_INTEL_GPU_CFL__ == 1) ||
+    (__SYCL_TARGET_INTEL_GPU_APL__ == 1) ||
+    (__SYCL_TARGET_INTEL_GPU_GLK__ == 1) ||
+    (__SYCL_TARGET_INTEL_GPU_WHL__ == 1) ||
+    (__SYCL_TARGET_INTEL_GPU_AML__ == 1) ||
+    (__SYCL_TARGET_INTEL_GPU_CML__ == 1) ||
+    (__SYCL_TARGET_INTEL_GPU_ICLLP__ == 1) ||
+    (__SYCL_TARGET_INTEL_GPU_EHL__ == 1) ||
+    (__SYCL_TARGET_INTEL_GPU_TGLLP__ == 1) ||
+    (__SYCL_TARGET_INTEL_GPU_RKL__ == 1) ||
+    (__SYCL_TARGET_INTEL_GPU_ADL_S__ == 1) ||
+    (__SYCL_TARGET_INTEL_GPU_RPL_S__ == 1) ||
+    (__SYCL_TARGET_INTEL_GPU_ADL_P__ == 1) ||
+    (__SYCL_TARGET_INTEL_GPU_ADL_N__ == 1) ||
+    (__SYCL_TARGET_INTEL_GPU_DG1__ == 1) ||
+    (__SYCL_TARGET_INTEL_GPU_ACM_G10__ == 1) ||
+    (__SYCL_TARGET_INTEL_GPU_ACM_G11__ == 1) ||
+    (__SYCL_TARGET_INTEL_GPU_ACM_G12__ == 1) ||
+    (__SYCL_TARGET_INTEL_GPU_PVC__ == 1);
+
+// One entry for each enumerator in "architecture" telling whether the AOT
+// target matches that architecture.
+static constexpr bool is_aot_for_architecture[] = {
+    (__SYCL_TARGET_INTEL_X86_64__ == 1),
+    (__SYCL_TARGET_INTEL_GPU_BDW__ == 1),
+    (__SYCL_TARGET_INTEL_GPU_SKL__ == 1),
+    (__SYCL_TARGET_INTEL_GPU_KBL__ == 1),
+    (__SYCL_TARGET_INTEL_GPU_CFL__ == 1),
+    (__SYCL_TARGET_INTEL_GPU_APL__ == 1),
+    (__SYCL_TARGET_INTEL_GPU_GLK__ == 1),
+    (__SYCL_TARGET_INTEL_GPU_WHL__ == 1),
+    (__SYCL_TARGET_INTEL_GPU_AML__ == 1),
+    (__SYCL_TARGET_INTEL_GPU_CML__ == 1),
+    (__SYCL_TARGET_INTEL_GPU_ICLLP__ == 1),
+    (__SYCL_TARGET_INTEL_GPU_EHL__ == 1),
+    (__SYCL_TARGET_INTEL_GPU_TGLLP__ == 1),
+    (__SYCL_TARGET_INTEL_GPU_RKL__ == 1),
+    (__SYCL_TARGET_INTEL_GPU_ADL_S__ == 1),
+    (__SYCL_TARGET_INTEL_GPU_RPL_S__ == 1),
+    (__SYCL_TARGET_INTEL_GPU_ADL_P__ == 1),
+    (__SYCL_TARGET_INTEL_GPU_ADL_N__ == 1),
+    (__SYCL_TARGET_INTEL_GPU_DG1__ == 1),
+    (__SYCL_TARGET_INTEL_GPU_ACM_G10__ == 1),
+    (__SYCL_TARGET_INTEL_GPU_ACM_G11__ == 1),
+    (__SYCL_TARGET_INTEL_GPU_ACM_G12__ == 1),
+    (__SYCL_TARGET_INTEL_GPU_PVC__ == 1)};
+
+// Read the value of "is_allowable_aot_mode" via a template to defer triggering
+// static_assert() until template instantiation time.
+template <ext::intel::experimental::architecture... Archs>
+constexpr static bool allowable_aot_mode() {
+  return is_allowable_aot_mode;
+}
+
+// Tells if the current device has one of the architectures in the parameter
+// pack.
+template <ext::intel::experimental::architecture... Archs>
+constexpr static bool device_architecture_is() {
+  return (is_aot_for_architecture[static_cast<int>(Archs)] || ...);
```

Comment:
That seems a bit fragile: I mean that incorrect order of initialization of `is_aot_for_architecture` could lead to incorrect results returned here. Unfortunately, I don't have any good suggestions about how to make it more reliable by at least adding some self-correctness checks.

We should at least add a comment around both `architecture` enumerator and `is_aot_for_architecture` variable that the order must strictly match between those two.

Reply:
I tried to make it more explicit but stuck of course with detecting the size of enum, you can see some draft here: https://godbolt.org/z/G9PYTaxcr . So I will add comments to the code.

Reply:
> I tried to make it more explicit but stuck of course with detecting the size of enum

Not sure that it is a good way to do with a "public", i.e. exposed to user `enum`, but we could add some "internal" enumerator in there at the end, which would represent the size. Something like `__last = 23`, or even without `= 23` part to let the compiler calculate that for us automatically

Reply:
I also think it is not good to expose one more item in this enum which is not a name of architecture and which is not documented in spec. But thought about adding something like `SIZE,` after `intel_gpu_pvc,` so we could write bool `arr[static_cast<int>(architecture::SIZE)];`
@gmlueck what do you think about adding an item to ```architecture``` enum which will basically count the number of items in this enum? The purpose is to make the code safer after adding more architectures to this extension in the future.

Reply:
This seems like a general problem that probably comes up with other enums.  We don't have a common pattern for solving this?  It would be nice to avoid adding a publicly visible enumerator for this.  Could we use a pattern like:

```
enum class architecture {
  foo,
  bar
  // Update "max_architecture" below if you add new elements here!
};

namespace detail {
static constexpr architecture max_architecture = architecture::bar;
}
```

Reply:
We can probably use the idea from https://taylorconor.com/blog/enum-reflection/ to do a `static_assert` (without introducing the dependency onto https://github.com/Neargye/magic_enum).

Reply:
Or the `static_assert` (and all the C++ metaprogramming) could be in a unit test, keeping it entirely separate from the public header.

Reply:
Thanks everyone, made as short as possible using @gmlueck suggestion. Now if more architectures will be added (by updating `architecture` enum and `IsAOTForArchitectureClass::arr`) and developer forgets to increase `max_architecture`, there will be a compile error.

Reply:
I must be blind in the morning. Where is this tested?